geps/gep-1897.md

```diff
@@ -120,12 +137,335 @@ was followed by intense discussion and closed in favor of a downsize in scope.
+- An indication that system certificates may be used.
+
+TLSConnectionPolicy is defined as a Direct Policy Attachment without defaults or overrides, applied to a Service that
+accesses the backend in question, where the TLSConnectionPolicy resides in the same namespace as the Service it is
+applied to.  The TLSConnectionPolicy and the Service must reside in the same namespace in order to prevent the
+complications involved with sharing trust across namespace boundaries.  We chose the Service resource as a target,
+rather than the Route resource, so that we can reuse the same TLSConnectionPolicy for all the different Routes that
+might point to this Service.
+For the use case where certificates are stored in their own namespace, users may create Secrets and use ReferenceGrants
+for a TLSConnectionPolicy-to-Secret binding.  Implementations must respect a ReferenceGrant for cross-namespace Secret
+sharing to TLSConnectionPolicy, even if they don't for other cross-namespace sharing.
+
+One of the areas of concern for this API is that we need to indicate how and when the API implementations should use the
+backend destination certificate authority.  This solution proposes, as introduced in
+[GEP-713](https://gateway-api.sigs.k8s.io/geps/gep-713/), that the implementation
+should watch the connections to a specified TargetRef (Service), and if the Service matches a TLSConnectionPolicy, then
+assume the connection is TLS, and verify that the TargetRef’s certificate can be validated by the client (Gateway) using
+the provided certificates and hostname before the connection is made. On the question of how to signal
+that there was a failure in the certificate validation, this is left up to the implementation to return a response error
+that is appropriate, such as one of the HTTP error codes: 400 (Bad Request), 401 (Unauthorized), 403 (Forbidden), or
+other signal that makes the failure sufficiently clear to the requester without revealing too much about the transaction,
+based on established security requirements.
+
+All policy resources must include `TargetRef` with the fields specified
+[here](https://github.com/kubernetes-sigs/gateway-api/blob/a33a934af9ec6997b34fd9b00d2ecd13d143e48b/apis/v1alpha2/policy_types.go#L24-L41).
+In an upcoming [extension](https://github.com/kubernetes-sigs/gateway-api/issues/2147) to TargetRef, policy resources
+_may_ also choose to include `SectionName` and/or `Port` in the TargetRef following the same mechanics as `ParentRef`.
+
+TLSConnectionPolicySpec contains the `TargetRef` and `TLS` fields.  The `TLS` field is a `TLSConnectionPolicyConfig` and
+contains `CertRefs`, `CACertRefs`, and  `Hostname`.
+The names of the fields were chosen to facilitate discussion, but may be substituted without blocking acceptance of the
+content of the API change.
+
+The `TrustedCA` contains the `CertRefs` and `CACertRefs` fields. CertRefs is a slice of
```

Comment:
I know, naming is hard, but I think it's important that the field currently called `CACertRefs` does _not_ end with `Refs` - that is reserved in the API conventions for structs that contain a list of references to objects.

How about instead of `CACertRefs`, we have `UseExtraCerts` - I suggest this because it has two possible values (YAML shown for ease of typing):
- `trustedCAs.useExtraCerts: ""` (or unspecified) - don't use any extra certificates for Trusted CAs.
- `trustedCAs.useExtraCerts: "system"` - use the system certificates as extra trusted CAs.

This name _implies_ that it's possible to use `CertRefs` _with_ `UseExtraCerts` - which I think has undefined behavior at this time.

Other names I can think of:
- `AdditionalCerts` (YAML `trustedCAs.additionalCerts`) Also implies a merging behavior between `CertRefs` and `AdditionalCerts`.
- `AlternativeCerts` (YAML `trustedCAs.alternativeCerts`) Implies that you can only choose _either_ `CertRefs` _or_ `AdditionalCerts` - they are mutually exclusive.

Whatever name we choose, we need to ensure that the behavior when you set both is defined. ("That's a syntax error" is okay, as is "the lists must be merged" or other possible answers, we just need to pick one, which we can also mark as temporary - given that this is `Provisional`.)

Again, naming is hard. ping @robscott as well as @candita.

Reply:
Yeah I've struggled with naming on this one as well. Of course I'd originally proposed `wellKnownRoots` but obviously that has it's flaws, and not all systems are necessarily going to be using "well-known" values. So _maybe_ we should just narrow down the scope of this field a bit. It seems _very unlikely_ that Kubernetes or Gateway API will ever try to maintain a common set of certificates that should be used for this purpose, so we can likely say that whatever values are supported here other than none/empty, are going to be at least somewhat implementation-specific. 

I think we should use the following guidelines when choosing a name here:

1. Avoid a name that locks us into whether/how this combines with `certRefs`. My personal opinion is that we should start with validation that ensures that only `certRefs` or this field can be set, and we can loosen that later if needed. I think this would rule out `additional`, `alternative`, or `extra`.
2. Avoid a boolean value so we're not conflicting with the [relevant Kubernetes API Conventions](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#primitive-types).
3. Building on 2, allow for some theoretical room for expansion. I think the most likely value here would be domain-prefixed values if for some reasons implementations want to support predefined sets of CAs. Even this seems highly unlikely at this point, but want to at least leave room for it so we're not boxed in years from now.

With all that said, I'm struggling to actually find a good name that meets this criteria. Some more ideas:

A. `trustedCAs.systemCerts: "System"`
B. `trustedCAs.predefinedCerts: "System"`
C. `trustedCAs.standardCerts: "System"`

As far as whether the empty value should be explicit ("None") or just empty (nil pointer or ""), I tend to lean towards a default of "None" so there's absolutely no ambiguity here. If we'd rather avoid "None", I think I'd lean towards a nil pointer instead of "", but open to what others think.

Reply:
@youngnick  I changed the name of `CACertRefs`, didn't realize the Refs suffix had a semantic association.  Should I also change the name of `CertRefs`, which a slice of ConfigMapNames, or is that fitting?

@robscott and I had this conversation last night and I originally proposed `alternativeCerts` (amongst others) but I get Rob's point about preferring to use a name that doesn't lock us into how this field _combines_ with `certRefs`.  I chose `standardCerts`.  `certRefs` and `standardCerts` are intended to be mutually exclusive in the first round.

Regarding using a nil pointer instead of "" for `standardCerts`, I'd rather leave it "" so that the field is visible when yaml is examined.  If we don't normally do that for our users, I can make it a pointer.

Reply:
Oh, I missed that `CertRefs` is a slice of strings that are ConfigMap names. Yes, that should be a `[]ObjectReference` instead, with a default of Configmap as the Kind (which puts "", meaning "core" as the Group). That way, you can specify a list of certificates in Configmaps like this:
```yaml
    ...
    certRefs:
    - name: CAcert1
    - name: CAcert2
    ...
```
or, you can also specify other objects like Secrets if you want. Anything _not_ a ConfigMap should be Implementation Specific support in the initial version though. This lets us change that _much_ more easily lately.  (Practically, to be able to set the defaults, you'll need to create another type like `SecretObjectReference` - probably `ConfigmapObjectReference`).

Using a nil pointer or "" as the default value makes no difference - because the field is optional and will have `omitempty` set, then a nil pointer (for a `*string` type) or an empty string (for a `string` type) will both be not necessary in input YAMLs, and hidden in output YAMLs. The only way you'll see the `standardCerts` field (which I agree is a good name) is if it's set to `system`. However, if you _do_ set `standardCerts: system`, then the `CertRefs` field also must be empty in this version, so it will also be hidden on output.

So you'll only ever see:
```yaml
    ...
    certRefs:
    - name: CAcert1
    ...
```
or
```yaml
    ...
	standardCerts: "system"
    ...
```
Which I think is what we want.

So, my requested changes here are:
- `CertRefs` should be a `[]ConfigmapObjectReference`, where that is an ObjectReference struct that defaults to Configmap.
- `standardCerts` should be a plain string field (not a `*string`), with a default value of "" that's optional.

Reply:
@youngnick we discussed offline, and I will make the requested change to `CertRefs` and do `StandardCerts` like this:

* standardCerts should be a `StandardCertType` pointer field with Enum validation, and one valid value: System.  An informal review showed that alias strings usually start with a capital letter, particularly condition types and reasons, so I used System rather than system.